processing/src/main/java/io/druid/segment/data/ShapeShiftingColumnarIntsSerializer.java

```diff
@@ -0,0 +1,198 @@
+        intBlockSize,
+        byteOrder
+    );
+    final CompressibleIntFormEncoder bytepack = new BytePackedIntFormEncoder(intBlockSize, byteOrder);
+
+    final IntFormEncoder[] defaultCodecs = new IntFormEncoder[]{
+        new ZeroIntFormEncoder(intBlockSize, byteOrder),
+        new ConstantIntFormEncoder(intBlockSize, byteOrder),
+        new UnencodedIntFormEncoder(intBlockSize, byteOrder),
+        rle,
+        bytepack,
+        new CompressedIntFormEncoder(
+            intBlockSize,
+            byteOrder,
+            compressionStrategy,
+            rle,
+            uncompressedDataBuffer,
+            compressedDataBuffer
+        ),
+        new CompressedIntFormEncoder(
+            intBlockSize,
+            byteOrder,
+            compressionStrategy,
+            bytepack,
+            uncompressedDataBuffer,
+            compressedDataBuffer
+        ),
+        new LemireIntFormEncoder(intBlockSize, IntCodecs.FASTPFOR, "fastpfor", byteOrder)
+    };
+
+    return defaultCodecs;
+  }
+
+  public ShapeShiftingColumnarIntsSerializer(
+      final SegmentWriteOutMedium segmentWriteOutMedium,
+      final IntFormEncoder[] codecs,
+      final IndexSpec.ShapeShiftOptimizationTarget optimizationTarget,
+      final IndexSpec.ShapeShiftBlockSize blockSize,
+      @Nullable final ByteOrder overrideByteOrder
+  )
+  {
+    this(
+        segmentWriteOutMedium,
+        codecs,
+        optimizationTarget,
+        blockSize,
+        overrideByteOrder,
+        null
+    );
+  }
+
+  @VisibleForTesting
+  public ShapeShiftingColumnarIntsSerializer(
+      final SegmentWriteOutMedium segmentWriteOutMedium,
+      final IntFormEncoder[] codecs,
+      final IndexSpec.ShapeShiftOptimizationTarget optimizationTarget,
+      final IndexSpec.ShapeShiftBlockSize blockSize,
+      @Nullable final ByteOrder overrideByteOrder,
+      @Nullable final Byte overrideLogValuesPerChunk
+  )
+  {
+    super(
+        segmentWriteOutMedium,
+        codecs,
+        optimizationTarget,
+        blockSize,
+        2,
+        ShapeShiftingColumnarInts.VERSION,
+        overrideByteOrder,
+        overrideLogValuesPerChunk
+    );
+
+    Closer closer = segmentWriteOutMedium.getCloser();
+    if (closer != null) {
+      closer.register(() -> {
+        if (unencodedValuesHolder != null) {
+          unencodedValuesHolder.close();
+        }
+      });
+    }
+
+    // enable optimization of encoders such as rle if there are additional non-zero and non-constant encoders.
+    this.enableEncoderOptOut =
+        Arrays.stream(codecs)
+              .filter(e -> !(e instanceof ZeroIntFormEncoder) && !(e instanceof ConstantIntFormEncoder))
+              .count() > 1;
+  }
+
+  @Override
+  public void initializeChunk()
+  {
+    unencodedValuesHolder = CompressedPools.getShapeshiftIntsDecodedValuesArray(logValuesPerChunk);
+    currentChunk = unencodedValuesHolder.get();
```

Comment:
we need just one int[] as things happen sequentially. would it not be simpler to allocate that array instead of using object pool?